src/Features/Core/Portable/InvertLogical/AbstractInvertLogicalCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InvertLogical
+{
+    /// <summary>
+    /// Code refactoring to help convert code like `!a || !b` to `!(a &amp;&amp; b)`
+    /// </summary>
+    internal abstract class AbstractInvertLogicalCodeRefactoringProvider<
+        TSyntaxKind, 
+        TExpressionSyntax,
+        TBinaryExpressionSyntax>
+        : CodeRefactoringProvider
+        where TSyntaxKind : struct
+        where TExpressionSyntax : SyntaxNode
+        where TBinaryExpressionSyntax : TExpressionSyntax
+    {
+        private static readonly SyntaxAnnotation s_annotation = new SyntaxAnnotation();
```

Comment:
>s_annotation [](start = 49, length = 12)

A comment would be useful (what's the purpose of this annotation?)

Reply:
Comment is actually deeper where it is used in the two pass form.  I originally had it here, but i moved it because i thought it was clearer later on.  I'll add a comment here saying there is a later comment :)